In [34]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.decomposition import PCA
import plotly.express as px
import tensorflow as tf
from tensorflow import keras
# SVD is used for dimensionality reduction



### Data Import

In [35]:
train_data = pd.read_csv('kaggle_train.csv')

train_data

Labels     ACAN      AGER      ALPK1    ANKRD17     APOB      APPL1  \
0         0  13.6584   88.0205   402.5420  1424.2625   3.7940  1001.2330   
1         0  22.0844   86.3807   690.4861  2226.3285  20.4071   486.9744   
2         0  32.7422   53.0820   128.4882  1073.0497   0.0000   654.8431   
3         0  22.7975   42.1175   634.0804  3285.5487   0.0000   574.9614   
4         0   2.5000  134.0625   532.1875   936.5625   0.0000  1933.4375   
..      ...      ...       ...        ...        ...      ...        ...   
247       1  61.3497  209.2706   300.6135   676.8916   4.0900   866.3940   
248       1  25.0447   89.7436  1050.3876  1677.6983   0.0000  1019.0817   
249       0  18.8321   50.5493   631.0399  1997.8525   0.3304   821.6734   
250       0  17.2414   92.8100   825.3852  2262.6559   0.3668  1421.8635   
251       0  30.6291   96.0265   541.6667  2443.4327   0.2759  2218.5403   

         APPL2      ARRB2    ASGR1  ...      UBE2N     UBE2V1     UBQLN1  \
0     508.0148   567.2010  13.6584  ...  2437.6363  3413.2107  4135.4453   
1    1448.0640   651.9083  11.1820  ...  2153.9253  1817.3762  3426.4325   
2     492.6206   512.9604  13.8906  ...  3122.4110  3720.6995  3000.3721   
3     513.9104  1033.2303  31.6847  ...  2663.4467  3217.9212  4131.3756   
4     679.3750  1175.6250   6.5625  ...  1898.4375  3009.1375  2549.3750   
..         ...        ...      ...  ...        ...        ...        ...   
247   836.4008   903.8855  40.8998  ...  2700.7498  2858.8957  1663.9400   
248  2058.1395  2086.4639   3.8760  ...  1367.0244  2176.7352  2595.4085   
249   941.9344   912.1996  17.8409  ...  2085.7355  2405.6562  3461.1382   
250   960.7483  1231.1079   2.9347  ...  2774.0279  2888.9142  4561.6288   
251  1235.9272   961.6446   4.4150  ...  2553.8079  2839.9558  4388.2450   

          UFD1    UNC93B1  USP17L2       VCAN      WDFY1       XIAP    ZCCHC3  
0     950.0104  1434.1269   0.0000    31.8695   923.0769  1046.0021  819.1217  
1    1380.1196   815.7241   0.0000  1651.0166  1473.7825  1618.8684  615.8465  
2    2681.3791  1429.2447   0.4961   693.5384   477.2417  1064.6161  804.1672  
3    1178.8640  1018.5471   0.0000   675.0386  1596.5997  1666.1515  402.6275  
4    1293.1156   744.6875   0.0000    69.0625  1651.5625  1250.3125  982.5000  
..         ...        ...      ...        ...        ...        ...       ...  
247  2168.3640  1858.2140   0.0000   441.0361   586.9121   583.5037  732.1063  
248  1222.4120  2648.4794   0.0000   471.9738  1548.5987  1888.4914  788.6106  
249  1596.0981   665.7306   0.3304  1272.3218   763.8556  1522.4250  347.2371  
250  1594.2296   653.7051   0.0000   392.5165  2422.9640  1749.0829  768.1585  
251  1104.2908   602.6490   0.0000   411.4238   835.5408  1103.2009  713.3002  

[252 rows x 319 columns]

In [36]:
x_train = train_data.drop('Labels', axis=1)

x_train

ACAN      AGER      ALPK1    ANKRD17     APOB      APPL1      APPL2  \
0    13.6584   88.0205   402.5420  1424.2625   3.7940  1001.2330   508.0148   
1    22.0844   86.3807   690.4861  2226.3285  20.4071   486.9744  1448.0640   
2    32.7422   53.0820   128.4882  1073.0497   0.0000   654.8431   492.6206   
3    22.7975   42.1175   634.0804  3285.5487   0.0000   574.9614   513.9104   
4     2.5000  134.0625   532.1875   936.5625   0.0000  1933.4375   679.3750   
..       ...       ...        ...        ...      ...        ...        ...   
247  61.3497  209.2706   300.6135   676.8916   4.0900   866.3940   836.4008   
248  25.0447   89.7436  1050.3876  1677.6983   0.0000  1019.0817  2058.1395   
249  18.8321   50.5493   631.0399  1997.8525   0.3304   821.6734   941.9344   
250  17.2414   92.8100   825.3852  2262.6559   0.3668  1421.8635   960.7483   
251  30.6291   96.0265   541.6667  2443.4327   0.2759  2218.5403  1235.9272   

         ARRB2    ASGR1    ASGR2  ...      UBE2N     UBE2V1     UBQLN1  \
0     567.2010  13.6584   1.8970  ...  2437.6363  3413.2107  4135.4453   
1     651.9083  11.1820   9.5047  ...  2153.9253  1817.3762  3426.4325   
2     512.9604  13.8906   1.4883  ...  3122.4110  3720.6995  3000.3721   
3    1033.2303  31.6847   3.8640  ...  2663.4467  3217.9212  4131.3756   
4    1175.6250   6.5625  18.7500  ...  1898.4375  3009.1375  2549.3750   
..         ...      ...      ...  ...        ...        ...        ...   
247   903.8855  40.8998  23.1766  ...  2700.7498  2858.8957  1663.9400   
248  2086.4639   3.8760   2.9815  ...  1367.0244  2176.7352  2595.4085   
249   912.1996  17.8409  12.2243  ...  2085.7355  2405.6562  3461.1382   
250  1231.1079   2.9347   5.8694  ...  2774.0279  2888.9142  4561.6288   
251   961.6446   4.4150  27.3179  ...  2553.8079  2839.9558  4388.2450   

          UFD1    UNC93B1  USP17L2       VCAN      WDFY1       XIAP    ZCCHC3  
0     950.0104  1434.1269   0.0000    31.8695   923.0769  1046.0021  819.1217  
1    1380.1196   815.7241   0.0000  1651.0166  1473.7825  1618.8684  615.8465  
2    2681.3791  1429.2447   0.4961   693.5384   477.2417  1064.6161  804.1672  
3    1178.8640  1018.5471   0.0000   675.0386  1596.5997  1666.1515  402.6275  
4    1293.1156   744.6875   0.0000    69.0625  1651.5625  1250.3125  982.5000  
..         ...        ...      ...        ...        ...        ...       ...  
247  2168.3640  1858.2140   0.0000   441.0361   586.9121   583.5037  732.1063  
248  1222.4120  2648.4794   0.0000   471.9738  1548.5987  1888.4914  788.6106  
249  1596.0981   665.7306   0.3304  1272.3218   763.8556  1522.4250  347.2371  
250  1594.2296   653.7051   0.0000   392.5165  2422.9640  1749.0829  768.1585  
251  1104.2908   602.6490   0.0000   411.4238   835.5408  1103.2009  713.3002  

[252 rows x 318 columns]

In [37]:
y_train = train_data['Labels']

y_train

0      0
1      0
2      0
3      0
4      0
      ..
247    1
248    1
249    0
250    0
251    0
Name: Labels, Length: 252, dtype: int64

# Data Preprocessing

In [38]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252 entries, 0 to 251
Columns: 319 entries, Labels to ZCCHC3
dtypes: float64(318), int64(1)
memory usage: 628.2 KB


In [39]:
train_data.describe()

Labels        ACAN        AGER        ALPK1      ANKRD17  \
count  252.000000  252.000000  252.000000   252.000000   252.000000   
mean     0.507937   48.349698   87.582435   458.638724  1735.596670   
std      0.500932   86.843637   62.492394   256.577766   683.408076   
min      0.000000    0.395900    9.856400    44.263900   148.780500   
25%      0.000000    8.124100   46.051875   274.031125  1324.797475   
50%      1.000000   21.252850   73.756000   403.065750  1683.823550   
75%      1.000000   52.228225  105.536225   627.079150  2159.678675   
max      1.000000  695.593700  405.115100  1807.708600  4201.452400   

             APOB        APPL1        APPL2        ARRB2       ASGR1  ...  \
count  252.000000   252.000000   252.000000   252.000000  252.000000  ...   
mean    10.103969  1048.303742   926.815081   959.023516   19.846768  ...   
std     48.803227   435.198678   523.033512   497.399230   26.660277  ...   
min      0.000000   271.750400   251.252100   261.989300    0.571800  ...   
25%      0.000000   745.653375   633.514450   612.724975    6.864525  ...   
50%      0.439400  1013.678150   794.230250   845.650200   13.430150  ...   
75%      2.302925  1312.827725  1041.623100  1157.327775   21.343450  ...   
max    575.249700  2790.092700  4749.258600  3152.486600  264.796600  ...   

             UBE2N        UBE2V1       UBQLN1         UFD1      UNC93B1  \
count   252.000000    252.000000   252.000000   252.000000   252.000000   
mean   2158.231899   2968.174126  3012.822705  1545.571726  1392.881560   
std     526.403209   1031.137240   853.086502   505.267660   977.981117   
min     766.683600   1177.463400  1071.951200   566.360100   239.984400   
25%    1834.701450   2432.261775  2440.509825  1184.313250   758.756525   
50%    2130.631800   2810.698300  2922.025050  1447.979150  1168.462300   
75%    2523.760700   3416.843800  3466.597375  1797.457550  1674.928150   
max    4123.979100  12055.091800  6188.692100  4349.902400  6831.658300   

          USP17L2          VCAN        WDFY1         XIAP       ZCCHC3  
count  252.000000    252.000000   252.000000   252.000000   252.000000  
mean     0.092321   1479.875324  1138.023175  1432.045627   675.118475  
std      0.249843   2745.942894   530.726910   604.559843   283.487914  
min      0.000000     18.094400   321.771000   293.969100   179.927400  
25%      0.000000    232.117175   735.068250  1048.561825   498.732950  
50%      0.000000    555.124750  1061.980400  1310.477400   621.197050  
75%      0.000000   1503.523350  1420.064175  1617.536325   794.773425  
max      2.056600  29131.572500  3511.057400  4605.813200  2616.190500  

[8 rows x 319 columns]

In [40]:
train_data['Labels'].value_counts()

1    128
0    124
Name: Labels, dtype: int64

Very minute class imbalance, so no need for oversampling

In [41]:
train_data.isnull().sum()

Labels     0
ACAN       0
AGER       0
ALPK1      0
ANKRD17    0
          ..
USP17L2    0
VCAN       0
WDFY1      0
XIAP       0
ZCCHC3     0
Length: 319, dtype: int64

There is no null value, so no need for data imputation

### Outliner Analysis

In [42]:
z_scores = (x_train - x_train.mean())/x_train.std()

z_scores

ACAN      AGER     ALPK1   ANKRD17      APOB     APPL1     APPL2  \
0   -0.399469  0.007010 -0.218634 -0.455561 -0.129294 -0.108159 -0.800714   
1   -0.302444 -0.019230  0.903614  0.718066  0.211116 -1.289823  0.996588   
2   -0.179720 -0.552074 -1.286746 -0.969475 -0.207035 -0.904094 -0.830147   
3   -0.294232 -0.727527  0.683776  2.267974 -0.207035 -1.087647 -0.789442   
4   -0.527957  0.743772  0.286653 -1.169190 -0.207035  2.033861 -0.473086   
..        ...       ...       ...       ...       ...       ...       ...   
247  0.149694  1.947248 -0.615896 -1.549155 -0.123229 -0.417992 -0.172865   
248 -0.268356  0.034583  2.306314 -0.084720 -0.207035 -0.067146  2.163006   
249 -0.339894 -0.592602  0.671926  0.383747 -0.200265 -0.520751  0.028907   
250 -0.358210  0.083651  1.429377  0.771222 -0.199519  0.858366  0.064878   
251 -0.204052  0.135121  0.323598  1.035744 -0.201382  2.688971  0.590999   

        ARRB2     ASGR1     ASGR2  ...     UBE2N    UBE2V1    UBQLN1  \
0   -0.787743 -0.232119 -0.707881  ...  0.530780  0.431598  1.315954   
1   -0.617442 -0.325007  0.105316  ... -0.008181 -1.116047  0.484839   
2   -0.896791 -0.223410 -0.751568  ...  1.831636  0.729801 -0.014595   
3    0.149190  0.444029 -0.497626  ...  0.959749  0.242205  1.311183   
4    0.435468 -0.498279  1.093560  ... -0.493527  0.039726 -0.543260   
..        ...       ...       ...  ...       ...       ...       ...   
247 -0.110853  0.789678  1.566725  ...  1.030613 -0.105979 -1.581179   
248  2.266671 -0.599047 -0.591958  ... -1.503045 -0.767540 -0.489299   
249 -0.094137 -0.075238  0.396018  ... -0.137720 -0.545532  0.525522   
250  0.547014 -0.634355 -0.283266  ...  1.169818 -0.076867  1.815532   
251  0.005270 -0.578830  2.009395  ...  0.751470 -0.124347  1.612289   

         UFD1   UNC93B1   USP17L2      VCAN     WDFY1      XIAP    ZCCHC3  
0   -1.178705  0.042174 -0.369514 -0.527326 -0.405004 -0.638553  0.507970  
1   -0.327454 -0.590152 -0.369514  0.062325  0.632640  0.309023 -0.209081  
2    2.247932  0.037182  1.616130 -0.286363 -1.245050 -0.607764  0.455218  
3   -0.725769 -0.382762 -0.369514 -0.293100  0.864054  0.387234 -0.961208  
4   -0.499648 -0.662788 -0.369514 -0.513781  0.967615 -0.300604  1.084284  
..        ...       ...       ...       ...       ...       ...       ...  
247  1.232599  0.475809 -0.369514 -0.378318 -1.038408 -1.403570  0.201024  
248 -0.639581  1.283867 -0.369514 -0.367051  0.773610  0.755005  0.400342  
249  0.099999 -0.743522  0.952914 -0.075586 -0.705010  0.149496 -1.156597  
250  0.096301 -0.755819 -0.369514 -0.395987  2.421096  0.524410  0.328198  
251 -0.873361 -0.808024 -0.369514 -0.389102 -0.569940 -0.543941  0.134686  

[252 rows x 318 columns]

In [43]:
z_scores.describe()


ACAN          AGER         ALPK1       ANKRD17          APOB  \
count  2.520000e+02  2.520000e+02  2.520000e+02  2.520000e+02  2.520000e+02   
mean  -6.167906e-17 -2.819614e-17  2.114711e-17 -6.344132e-17  2.202823e-17   
std    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min   -5.521855e-01 -1.243768e+00 -1.615007e+00 -2.321916e+00 -2.070348e-01   
25%   -4.631957e-01 -6.645698e-01 -7.194996e-01 -6.011038e-01 -2.070348e-01   
50%   -3.120188e-01 -2.212499e-01 -2.165931e-01 -7.575725e-02 -1.980313e-01   
75%    4.466104e-02  2.872956e-01  6.564888e-01  6.205399e-01 -1.598469e-01   
max    7.452981e+00  5.081141e+00  5.257938e+00  3.608175e+00  1.158009e+01   

              APPL1         APPL2       ARRB2         ASGR1         ASGR2  \
count  2.520000e+02  2.520000e+02  252.000000  2.520000e+02  2.520000e+02   
mean  -8.458842e-17  1.674146e-16    0.000000  4.229421e-17  8.635068e-17   
std    1.000000e+00  1.000000e+00    1.000000  1.000000e+00  1.000000e+00   
min   -1.784365e+00 -1.291625e+00   -1.401358 -7.229845e-01 -9.106545e-01   
25%   -6.954303e-01 -5.607683e-01   -0.696218 -4.869508e-01 -6.372236e-01   
50%   -7.956272e-02 -2.534920e-01   -0.227932 -2.406809e-01 -3.896871e-01   
75%    6.078235e-01  2.195041e-01    0.398682  5.613902e-02  3.960183e-01   
max    4.002285e+00  7.308219e+00    4.409864  9.187820e+00  6.933957e+00   

       ...         UBE2N        UBE2V1        UBQLN1          UFD1  \
count  ...  2.520000e+02  2.520000e+02  2.520000e+02  2.520000e+02   
mean   ... -4.229421e-17 -8.811294e-17  1.339317e-16  2.114711e-17   
std    ...  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min    ... -2.643503e+00 -1.736637e+00 -2.275117e+00 -1.938006e+00   
25%    ... -6.146058e-01 -5.197294e-01 -6.708732e-01 -7.149844e-01   
50%    ... -5.243148e-02 -1.527205e-01 -1.064343e-01 -1.931503e-01   
75%    ...  6.943894e-01  4.351212e-01  5.319211e-01  4.985196e-01   
max    ...  3.734299e+00  8.812520e+00  3.722799e+00  5.550188e+00   

            UNC93B1       USP17L2        VCAN         WDFY1          XIAP  \
count  2.520000e+02  2.520000e+02  252.000000  2.520000e+02  2.520000e+02   
mean   1.251204e-16 -2.114711e-17    0.000000  1.621278e-16 -2.819614e-17   
std    1.000000e+00  1.000000e+00    1.000000  1.000000e+00  1.000000e+00   
min   -1.178854e+00 -3.695140e-01   -0.532342 -1.537989e+00 -1.882488e+00   
25%   -6.484021e-01 -3.695140e-01   -0.454401 -7.592510e-01 -6.343190e-01   
50%   -2.294720e-01 -3.695140e-01   -0.336770 -1.432804e-01 -2.010855e-01   
75%    2.883968e-01 -3.695140e-01    0.008612  5.314240e-01  3.068194e-01   
max    5.561229e+00  7.862042e+00   10.070019  4.471291e+00  5.249716e+00   

             ZCCHC3  
count  2.520000e+02  
mean   1.762259e-17  
std    1.000000e+00  
min   -1.746780e+00  
25%   -6.221977e-01  
50%   -1.902071e-01  
75%    4.220813e-01  
max    6.847107e+00  

[8 rows x 318 columns]

In [44]:
fig = px.box(x_train, y=x_train.columns, title="Boxplot of Z-Scores of Features")
fig.show()


In [45]:
# for each column make a list of rows where the z-score is > 3 or < -3

outlier_rows = []
for col in z_scores.columns:
    outlier_rows.append(z_scores[(z_scores[col] > 1.5) | (z_scores[col] < -1.5)].index.tolist())
    
print(outlier_rows)

[[9, 17, 25, 50, 100, 106, 172, 179, 244], [14, 20, 29, 58, 75, 78, 129, 145, 161, 164, 166, 167, 176, 212, 220, 233, 247], [21, 53, 77, 81, 90, 92, 102, 105, 130, 136, 142, 154, 157, 159, 167, 185, 197, 200, 206, 234, 235, 248], [3, 5, 7, 20, 30, 33, 43, 51, 56, 82, 84, 92, 100, 105, 109, 114, 119, 120, 131, 132, 151, 162, 165, 169, 170, 176, 198, 199, 206, 226, 232, 247], [6, 59, 68, 104, 139, 209, 223], [4, 7, 24, 43, 55, 56, 61, 66, 80, 102, 105, 106, 109, 113, 114, 118, 122, 123, 131, 146, 152, 175, 192, 200, 232, 241, 251], [21, 31, 52, 53, 68, 89, 107, 173, 175, 179, 185, 211, 222, 224, 248], [8, 11, 21, 40, 52, 70, 85, 87, 96, 131, 150, 157, 175, 183, 186, 192, 196, 206, 215, 231, 236, 248], [18, 49, 94, 131, 165, 166, 193, 200, 215, 222, 245], [5, 20, 24, 30, 40, 43, 59, 68, 76, 85, 111, 114, 120, 122, 128, 136, 195, 215, 219, 247, 251], [11, 27, 33, 49, 53, 54, 74, 78, 86, 88, 97, 104, 113, 123, 129, 173, 178, 182, 194, 204, 206, 207, 213, 222], [1, 17, 31, 56, 179, 244], [16

In [46]:
# make a list of elements which are present in more than 50% sublists in outliner_rows list

sus_rows = []

for i in range (1, 252):
    sum_i = 0
    for list in outlier_rows:
        if i in list:
            sum_i += 1
    if sum_i >= len(outlier_rows) * 0.15:
        sus_rows.append(i)

print(sus_rows)
                


[40, 43, 95, 131, 136, 170, 195, 215]


In [47]:
# remove sus_rows from x_train and y_train

x_train = x_train.drop(sus_rows, axis=0)
y_train = y_train.drop(sus_rows, axis=0)

We removed those data points who apeared in more than 15% (=48 features / 318 features) as a outliner

In [48]:
x_train

ACAN      AGER      ALPK1    ANKRD17     APOB      APPL1      APPL2  \
0    13.6584   88.0205   402.5420  1424.2625   3.7940  1001.2330   508.0148   
1    22.0844   86.3807   690.4861  2226.3285  20.4071   486.9744  1448.0640   
2    32.7422   53.0820   128.4882  1073.0497   0.0000   654.8431   492.6206   
3    22.7975   42.1175   634.0804  3285.5487   0.0000   574.9614   513.9104   
4     2.5000  134.0625   532.1875   936.5625   0.0000  1933.4375   679.3750   
..       ...       ...        ...        ...      ...        ...        ...   
247  61.3497  209.2706   300.6135   676.8916   4.0900   866.3940   836.4008   
248  25.0447   89.7436  1050.3876  1677.6983   0.0000  1019.0817  2058.1395   
249  18.8321   50.5493   631.0399  1997.8525   0.3304   821.6734   941.9344   
250  17.2414   92.8100   825.3852  2262.6559   0.3668  1421.8635   960.7483   
251  30.6291   96.0265   541.6667  2443.4327   0.2759  2218.5403  1235.9272   

         ARRB2    ASGR1    ASGR2  ...      UBE2N     UBE2V1     UBQLN1  \
0     567.2010  13.6584   1.8970  ...  2437.6363  3413.2107  4135.4453   
1     651.9083  11.1820   9.5047  ...  2153.9253  1817.3762  3426.4325   
2     512.9604  13.8906   1.4883  ...  3122.4110  3720.6995  3000.3721   
3    1033.2303  31.6847   3.8640  ...  2663.4467  3217.9212  4131.3756   
4    1175.6250   6.5625  18.7500  ...  1898.4375  3009.1375  2549.3750   
..         ...      ...      ...  ...        ...        ...        ...   
247   903.8855  40.8998  23.1766  ...  2700.7498  2858.8957  1663.9400   
248  2086.4639   3.8760   2.9815  ...  1367.0244  2176.7352  2595.4085   
249   912.1996  17.8409  12.2243  ...  2085.7355  2405.6562  3461.1382   
250  1231.1079   2.9347   5.8694  ...  2774.0279  2888.9142  4561.6288   
251   961.6446   4.4150  27.3179  ...  2553.8079  2839.9558  4388.2450   

          UFD1    UNC93B1  USP17L2       VCAN      WDFY1       XIAP    ZCCHC3  
0     950.0104  1434.1269   0.0000    31.8695   923.0769  1046.0021  819.1217  
1    1380.1196   815.7241   0.0000  1651.0166  1473.7825  1618.8684  615.8465  
2    2681.3791  1429.2447   0.4961   693.5384   477.2417  1064.6161  804.1672  
3    1178.8640  1018.5471   0.0000   675.0386  1596.5997  1666.1515  402.6275  
4    1293.1156   744.6875   0.0000    69.0625  1651.5625  1250.3125  982.5000  
..         ...        ...      ...        ...        ...        ...       ...  
247  2168.3640  1858.2140   0.0000   441.0361   586.9121   583.5037  732.1063  
248  1222.4120  2648.4794   0.0000   471.9738  1548.5987  1888.4914  788.6106  
249  1596.0981   665.7306   0.3304  1272.3218   763.8556  1522.4250  347.2371  
250  1594.2296   653.7051   0.0000   392.5165  2422.9640  1749.0829  768.1585  
251  1104.2908   602.6490   0.0000   411.4238   835.5408  1103.2009  713.3002  

[244 rows x 318 columns]

In [49]:
y_train = y_train.reset_index(drop=True)

y_train

0      0
1      0
2      0
3      0
4      0
      ..
239    1
240    1
241    0
242    0
243    0
Name: Labels, Length: 244, dtype: int64

### Data Normalization

In [50]:
normalizer = MinMaxScaler()

x_train = pd.DataFrame(normalizer.fit_transform(x_train))

x_train

0         1         2         3         4         5         6    \
0    0.019077  0.197754  0.203169  0.278879  0.006595  0.289668  0.057084   
1    0.031198  0.193606  0.366454  0.487142  0.035475  0.085463  0.266076   
2    0.046528  0.109360  0.047761  0.187683  0.000000  0.152121  0.053661   
3    0.032223  0.081620  0.334468  0.762178  0.000000  0.120401  0.058394   
4    0.003027  0.314240  0.276688  0.152243  0.000000  0.659834  0.095181   
..        ...       ...       ...       ...       ...       ...       ...   
239  0.087678  0.504516  0.145369  0.084817  0.007110  0.236125  0.130091   
240  0.035456  0.202114  0.570545  0.344685  0.000000  0.296755  0.401708   
241  0.026519  0.102953  0.332744  0.427816  0.000574  0.218367  0.153553   
242  0.024231  0.209872  0.442952  0.496575  0.000638  0.456695  0.157736   
243  0.043489  0.218009  0.282063  0.543515  0.000480  0.773044  0.218914   

          7         8         9    ...       308       309       310  \
0    0.105591  0.087122  0.034544  ...  0.497708  0.205536  0.569564   
1    0.134897  0.070636  0.173081  ...  0.413202  0.058828  0.420929   
2    0.086826  0.088668  0.027102  ...  0.701674  0.233804  0.331611   
3    0.266819  0.207129  0.070363  ...  0.564968  0.187583  0.568711   
4    0.316083  0.039882  0.341437  ...  0.337103  0.168389  0.237066   
..        ...       ...       ...  ...       ...       ...       ...   
239  0.222071  0.268477  0.422046  ...  0.576079  0.154577  0.051446   
240  0.631197  0.021997  0.054293  ...  0.178817  0.091865  0.246716   
241  0.224948  0.114966  0.222604  ...  0.392891  0.112910  0.428205   
242  0.335277  0.015731  0.106882  ...  0.597905  0.157337  0.658908   
243  0.242054  0.025585  0.497459  ...  0.532311  0.152836  0.622560   

          311       312       313       314       315       316       317  
0    0.141966  0.181159  0.000000  0.000473  0.188539  0.134203  0.262367  
1    0.301124  0.087343  0.000000  0.056088  0.361213  0.273532  0.178929  
2    0.782643  0.180419  0.241223  0.023200  0.048748  0.138730  0.256228  
3    0.226651  0.118113  0.000000  0.022565  0.399722  0.285032  0.091411  
4    0.268929  0.076567  0.000000  0.001751  0.416956  0.183894  0.329428  
..        ...       ...       ...       ...       ...       ...       ...  
239  0.592807  0.245496  0.000000  0.014527  0.083135  0.021717  0.226650  
240  0.242766  0.365384  0.000000  0.015590  0.384672  0.339109  0.249843  
241  0.381045  0.064588  0.160654  0.043081  0.138616  0.250076  0.068675  
242  0.380354  0.062764  0.000000  0.012861  0.658829  0.305203  0.241448  
243  0.199056  0.055019  0.000000  0.013510  0.161092  0.148115  0.218931  

[244 rows x 318 columns]

### Feature Analysis

In [51]:
pca = PCA(n_components=50)

In [52]:
x_train_pca = pca.fit_transform(x_train)

x_train_pca


array([[ 0.18042898,  0.17607089, -0.28994924, ...,  0.03784637,
        -0.00669974, -0.26943527],
       [ 0.08878239, -0.58266997, -0.04601713, ..., -0.04229183,
         0.34494724,  0.04129979],
       [-0.70588433,  0.60323736, -0.45262555, ..., -0.09170969,
        -0.15445206,  0.330744  ],
       ...,
       [ 1.07727953, -0.66887583, -0.59093325, ..., -0.22137676,
        -0.08507303,  0.1398524 ],
       [ 2.44534129, -0.13708234,  1.08409969, ..., -0.21554653,
        -0.09364758,  0.07107507],
       [ 0.38759529, -0.76396706, -0.11391335, ..., -0.04876883,
        -0.04584612,  0.05511387]])

In [53]:
x_train_pca = pd.DataFrame(x_train_pca)

x_train_pca

0         1         2         3         4         5         6   \
0    0.180429  0.176071 -0.289949  0.025477  0.746208 -0.398752  0.224989   
1    0.088782 -0.582670 -0.046017  0.178294 -0.251488  0.371548  0.492281   
2   -0.705884  0.603237 -0.452626 -0.332569  0.658135 -0.316983  0.998364   
3    1.956534 -0.686816 -0.473297  0.105399  0.371009  0.446189 -0.515666   
4    0.600043 -0.059019  0.166911  0.215475 -0.179447 -0.291962  0.299701   
..        ...       ...       ...       ...       ...       ...       ...   
239  0.530784  1.633961 -0.533294  0.162771  0.437402  0.486363  0.723236   
240 -0.090587  0.138311  1.315055  1.159499  0.046552  0.579951  0.192399   
241  1.077280 -0.668876 -0.590933  0.151244 -0.264052  0.468196 -0.520558   
242  2.445341 -0.137082  1.084100  0.024811 -0.485510 -1.233888  0.407246   
243  0.387595 -0.763967 -0.113913 -0.033283  0.271924  0.142265  0.207844   

           7         8         9   ...        40        41        42  \
0    0.176701 -0.077914  0.111418  ...  0.114512 -0.263285 -0.242493   
1    0.314334 -0.199878 -0.734577  ...  0.306643  0.118508  0.125196   
2    0.132407 -0.349517  0.012065  ... -0.036710 -0.176622  0.017472   
3   -0.350681 -0.782804  0.487755  ... -0.069351 -0.210898 -0.091213   
4   -0.073311  0.134981  0.884130  ...  0.012209 -0.028078 -0.050760   
..        ...       ...       ...  ...       ...       ...       ...   
239  0.241298 -0.046867 -0.121019  ... -0.019607 -0.048408 -0.076816   
240 -0.100283  0.077447 -0.202042  ...  0.177166 -0.032991 -0.132168   
241 -0.137453  0.253223 -0.376908  ...  0.140239  0.061979  0.187907   
242 -0.008378  0.125767 -0.104215  ... -0.262400  0.041306 -0.203348   
243  0.382551  1.411386 -0.151352  ...  0.073708 -0.217567 -0.017516   

           43        44        45        46        47        48        49  
0    0.112631 -0.050545 -0.067835 -0.146302  0.037846 -0.006700 -0.269435  
1    0.233856 -0.247814  0.246023 -0.025306 -0.042292  0.344947  0.041300  
2   -0.211219  0.052910 -0.181258  0.083847 -0.091710 -0.154452  0.330744  
3   -0.018429  0.154866  0.003781  0.173045 -0.211881 -0.319229 -0.025032  
4   -0.072782 -0.102888 -0.029146 -0.086803 -0.034330  0.270344 -0.141558  
..        ...       ...       ...       ...       ...       ...       ...  
239 -0.099027 -0.195048  0.462779  0.178096 -0.212070  0.015081  0.119653  
240  0.194653  0.168084  0.174890  0.163725  0.065907  0.142307  0.032157  
241  0.024358  0.098465 -0.243249 -0.131199 -0.221377 -0.085073  0.139852  
242 -0.081470  0.018559  0.217182  0.069469 -0.215547 -0.093648  0.071075  
243  0.114340 -0.196094  0.090550  0.167679 -0.048769 -0.045846  0.055114  

[244 rows x 50 columns]

In [54]:
x_train_pca.describe()

0             1             2             3             4   \
count  2.440000e+02  2.440000e+02  2.440000e+02  2.440000e+02  2.440000e+02   
mean  -2.912060e-17  5.460113e-18 -5.824121e-17  2.343299e-17  1.092023e-17   
std    8.016034e-01  6.698471e-01  5.614657e-01  4.879308e-01  4.369988e-01   
min   -1.344068e+00 -1.342354e+00 -1.257841e+00 -1.534543e+00 -1.127374e+00   
25%   -5.887775e-01 -4.431417e-01 -4.130514e-01 -2.936114e-01 -2.829949e-01   
50%   -1.725920e-01 -8.867218e-02 -8.325605e-02  5.546060e-02 -4.092206e-02   
75%    4.147607e-01  3.074194e-01  4.344063e-01  3.039612e-01  2.719239e-01   
max    2.449618e+00  2.548500e+00  1.315055e+00  1.600036e+00  1.615776e+00   

                 5             6             7             8             9   \
count  2.440000e+02  2.440000e+02  2.440000e+02  2.440000e+02  2.440000e+02   
mean   3.822079e-17  1.456030e-17 -4.732098e-17  1.615283e-17  1.592533e-18   
std    4.181881e-01  3.702297e-01  3.616629e-01  3.565911e-01  3.506183e-01   
min   -1.233888e+00 -8.330949e-01 -9.343961e-01 -1.127362e+00 -8.355592e-01   
25%   -2.494932e-01 -2.629284e-01 -2.173951e-01 -1.969980e-01 -2.240643e-01   
50%   -3.096505e-02 -6.640475e-03 -1.904793e-02  1.544531e-02 -1.814055e-02   
75%    2.993358e-01  2.162860e-01  1.931731e-01  1.817278e-01  2.083254e-01   
max    1.344996e+00  1.241313e+00  1.506143e+00  1.411386e+00  1.372468e+00   

       ...            40            41            42            43  \
count  ...  2.440000e+02  2.440000e+02  2.440000e+02  2.440000e+02   
mean   ...  3.640075e-18 -9.100189e-18  1.820038e-18 -2.366049e-17   
std    ...  1.885375e-01  1.880845e-01  1.859406e-01  1.842660e-01   
min    ... -5.405921e-01 -5.278373e-01 -4.756647e-01 -4.971163e-01   
25%    ... -1.082560e-01 -1.071892e-01 -1.090139e-01 -1.051267e-01   
50%    ...  5.541060e-03 -1.156919e-02 -1.515301e-02  7.963206e-03   
75%    ...  1.200967e-01  1.024854e-01  1.000820e-01  1.155349e-01   
max    ...  6.139601e-01  7.567542e-01  9.006539e-01  5.569277e-01   

                 44            45            46            47            48  \
count  2.440000e+02  2.440000e+02  2.440000e+02  2.440000e+02  2.440000e+02   
mean  -1.183025e-17 -1.683535e-17 -6.825142e-18  1.706285e-19 -3.139565e-17   
std    1.816076e-01  1.788428e-01  1.777312e-01  1.767085e-01  1.755751e-01   
min   -3.947125e-01 -4.279374e-01 -4.731495e-01 -5.127843e-01 -6.225857e-01   
25%   -1.305269e-01 -1.315895e-01 -1.216104e-01 -1.017100e-01 -1.133251e-01   
50%   -5.835561e-03 -1.266411e-02  7.875246e-04 -1.038308e-02 -2.518246e-04   
75%    9.846464e-02  1.123506e-01  1.138955e-01  1.127229e-01  1.074339e-01   
max    8.807984e-01  6.585744e-01  6.121371e-01  5.351911e-01  6.342160e-01   

                 49  
count  2.440000e+02  
mean  -8.463176e-17  
std    1.730752e-01  
min   -4.075548e-01  
25%   -1.221085e-01  
50%    2.215336e-03  
75%    8.608644e-02  
max    5.912320e-01  

[8 rows x 50 columns]

In [55]:
x_train_pca.isnull().sum()

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
16    0
17    0
18    0
19    0
20    0
21    0
22    0
23    0
24    0
25    0
26    0
27    0
28    0
29    0
30    0
31    0
32    0
33    0
34    0
35    0
36    0
37    0
38    0
39    0
40    0
41    0
42    0
43    0
44    0
45    0
46    0
47    0
48    0
49    0
dtype: int64

In [56]:
# number of 0 values in each column

x_train_pca.isin([0]).sum()

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
16    0
17    0
18    0
19    0
20    0
21    0
22    0
23    0
24    0
25    0
26    0
27    0
28    0
29    0
30    0
31    0
32    0
33    0
34    0
35    0
36    0
37    0
38    0
39    0
40    0
41    0
42    0
43    0
44    0
45    0
46    0
47    0
48    0
49    0
dtype: int64

# Model Training

In [57]:
# number of rows in x_train

len(x_train)

244

In [58]:
len(y_train)

244

In [59]:
model = keras.Sequential(
    [
        keras.Input(shape=(50,)),
        keras.layers.Dense(100, activation="relu"),
        keras.layers.Dropout(0.3),
        # keras.layers.Dense(100, activation="relu"),
        keras.layers.Dense(100, activation="relu", kernel_regularizer=keras.regularizers.l2(0.01)),
        keras.layers.Dense(1, activation="sigmoid", kernel_regularizer=keras.regularizers.l2(0.01))
    ]
)

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(x_train_pca, y_train, epochs=1000, batch_size=5)



In [60]:
# random forest

from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(n_estimators=100, max_depth=10, random_state=0)

rf.fit(x_train_pca, y_train)

rf.score(x_train_pca, y_train)

1.0

In [61]:
# plot loss and accuracy

# plt.plot(rf.history.history['loss'])

# Testing Model

### Importing Test Data

In [62]:
test = pd.read_csv('kaggle_test.csv')

In [63]:
# do same PCA and MinMaxScaler on test data

id_test = test['ID']

test = test.drop('ID', axis=1)

test = pd.DataFrame(normalizer.transform(test))

test


0         1         2         3         4         5         6    \
0    0.130798  0.310344  0.162915  0.339770  0.000000  0.316426  0.071150   
1    0.024743  0.438690  0.291813  0.511881  0.001055  0.260987  0.118437   
2    0.173382  0.536220  0.163078  0.140619  0.000000  0.014986  0.033845   
3    0.016826  0.080970  0.477615  0.484653  0.000000  0.206806  0.169151   
4    0.120349  0.062347  0.057563  0.237857  0.000845  0.433117  0.128869   
..        ...       ...       ...       ...       ...       ...       ...   
99   0.132779  0.085621  0.203763  0.428739  0.000000  0.439554  0.135044   
100  0.054146  0.138340  0.109896  0.346859  0.000000  0.183488  0.162969   
101  0.029954  0.117234  0.282641  0.399038  0.000000  0.390332  0.138353   
102  0.008952  0.171842  0.373260  0.279886  0.000000  0.239380  0.140789   
103  0.030294  0.091641  0.202487  0.317314  0.000000  0.570898  0.110394   

          7         8         9    ...       308       309       310  \
0    0.151479  0.058975  0.090383  ...  0.225946  0.197046  0.293320   
1    0.268968  0.177977  0.309393  ...  0.513211  0.085045  0.499064   
2    0.404273  0.328145  0.281792  ...  0.296065  0.335132  0.482154   
3    0.129812  0.020965  0.050819  ...  0.329394  0.107059  0.390426   
4    0.277513  0.565528  0.017696  ...  0.347381  0.187859  0.363722   
..        ...       ...       ...  ...       ...       ...       ...   
99   0.104277  0.075157  0.329670  ...  0.507416  0.066329  0.542740   
100  0.241607  0.018956  0.140091  ...  0.651938  0.154611  0.620281   
101  0.183303  0.056364  0.064403  ...  0.334756  0.176355  0.454341   
102  0.159695  0.007211  0.037669  ...  0.134791  0.003800  0.121701   
103  0.160772  0.120301  0.025621  ...  0.367251  0.137629  0.306021   

          311       312       313       314       315       316       317  
0    0.307443  0.019389  0.241321  0.023400  0.211297  0.176281  0.271265  
1    0.125885  0.143744  0.000000  0.066950  0.378756  0.143234  0.135613  
2    0.496444  0.226349  0.000000  0.040286  0.261755  0.052649  0.209153  
3    0.203134  0.080865  0.000000  0.001551  0.233076  0.216681  0.157534  
4    0.222855  0.110066  0.000000  0.039518  0.081480  0.093470  0.175656  
..        ...       ...       ...       ...       ...       ...       ...  
99   0.061736  0.054747  0.000000  0.256132  0.227461  0.453570  0.124220  
100  0.274213  0.061629  0.000000  0.028886  0.109906  0.085930  0.112979  
101  0.220411  0.101603  0.000000  0.035849  0.281197  0.315909  0.052604  
102  0.485906  0.200084  0.402314  0.000871  0.183066  0.555874  0.227738  
103  0.418701  0.163218  0.000000  0.030851  0.109429  0.149963  0.050022  

[104 rows x 318 columns]

In [64]:
test_pca = pca.transform(test)

test_pca

array([[-0.56691795, -0.19263886, -0.20678535, ..., -0.08248125,
        -0.11036661,  0.07227895],
       [ 1.19256453, -0.24480223, -0.31765684, ...,  0.10463908,
         0.0763766 , -0.0399992 ],
       [ 2.53587335,  1.44829079, -0.25666129, ..., -0.0375589 ,
        -0.14835298, -0.21590581],
       ...,
       [ 0.62008418, -0.44950298, -0.20189709, ..., -0.0693292 ,
         0.0068022 , -0.16225176],
       [-0.94624686,  0.62444888,  0.74002165, ..., -0.44434089,
         0.28676979, -0.09982261],
       [ 0.37820145,  0.07064911, -0.43924626, ..., -0.04512978,
         0.00549102, -0.10045107]])

In [65]:
test_pca = pd.DataFrame(test_pca)

test_pca

0         1         2         3         4         5         6   \
0   -0.566918 -0.192639 -0.206785 -0.307182 -0.128100 -0.184203 -0.338331   
1    1.192565 -0.244802 -0.317657  0.180366 -0.162609  0.256390 -0.205194   
2    2.535873  1.448291 -0.256661 -1.134967  0.628941  0.584017  0.211755   
3   -0.623826 -0.597472 -0.086957  0.257278 -0.152770 -0.360709 -0.249530   
4   -0.942909 -0.292798 -0.342864 -0.337287 -0.208517 -0.153403  0.189810   
..        ...       ...       ...       ...       ...       ...       ...   
99   0.476660 -1.195067  0.098391 -0.435782 -0.008000  1.013033  0.140338   
100  0.168183 -0.653469 -0.853505  0.148368  0.049400  0.316582  0.673038   
101  0.620084 -0.449503 -0.201897  0.221832  0.184843 -0.070010  0.097802   
102 -0.946247  0.624449  0.740022  0.526748 -0.042497  0.169111 -0.262706   
103  0.378201  0.070649 -0.439246  0.082578  0.683201 -0.166254 -0.629860   

           7         8         9   ...        40        41        42  \
0    0.022537  0.095977 -0.029891  ...  0.053692 -0.047845 -0.034226   
1    0.125868  0.024507 -0.544883  ... -0.037899 -0.065013 -0.137802   
2   -0.019098 -0.927993 -0.264260  ... -0.087389 -0.090573 -0.184284   
3    0.028266  0.044806  0.120414  ...  0.082139  0.099561 -0.056162   
4   -0.438366  0.183066 -0.424097  ... -0.045771 -0.020258  0.195229   
..        ...       ...       ...  ...       ...       ...       ...   
99  -0.117676  0.103065  0.445009  ... -0.223833  0.071619 -0.013744   
100  0.324634  0.386486  0.264598  ...  0.420271 -0.010365  0.173431   
101 -0.144503  0.192634  0.171918  ...  0.071383  0.073713 -0.116225   
102 -0.242261 -0.298356 -0.391964  ...  0.349551  0.189516  0.231336   
103 -0.020773  0.041904  0.604381  ...  0.156439  0.145449 -0.093028   

           43        44        45        46        47        48        49  
0    0.132900 -0.066357 -0.144287 -0.231662 -0.082481 -0.110367  0.072279  
1   -0.106929  0.085230  0.019716 -0.110876  0.104639  0.076377 -0.039999  
2    0.078794  0.114074 -0.092645 -0.243637 -0.037559 -0.148353 -0.215906  
3    0.087225  0.027208 -0.129002  0.091633  0.028433 -0.128494 -0.034637  
4   -0.261423 -0.026373  0.058453 -0.212252  0.157998 -0.207797  0.008823  
..        ...       ...       ...       ...       ...       ...       ...  
99   0.002042 -0.021242 -0.119180  0.012138  0.169758  0.283477  0.004424  
100  0.475418  0.260131  0.128900 -0.395754  0.216184 -0.151110  0.082332  
101 -0.006846  0.165073  0.031741 -0.052348 -0.069329  0.006802 -0.162252  
102  0.022864 -0.175572 -0.273618  0.159757 -0.444341  0.286770 -0.099823  
103 -0.265631  0.138973 -0.128811 -0.102336 -0.045130  0.005491 -0.100451  

[104 rows x 50 columns]

In [67]:
# rf.evaluate(test_pca)

predictions = rf.predict(test_pca)

predictions = pd.DataFrame(predictions)

predictions = predictions.round().astype(int)

In [68]:
output = pd.concat([id_test, predictions.round().astype(int)], axis=1)

output.columns = ['ID', 'Labels']

output

ID  Labels
0    1001       0
1    1002       0
2    1003       1
3    1004       0
4    1005       0
..    ...     ...
99   1100       0
100  1101       1
101  1102       0
102  1103       1
103  1104       0

[104 rows x 2 columns]

In [70]:
output.to_csv('output10.csv', index=False)